src/systems/filecoin_markets/storage_market/storage_market_actor.go

```diff
@@ -242,74 +235,299 @@ func (a *StorageMarketActorCode_I) PublishStorageDeals(rt Runtime, newStorageDea
+		st._unlockBalance(rt, newExpiredDeal.Deal().Proposal().Client(), newExpiredDeal.Deal().Proposal().TotalClientCollateral())
+
+	}
+}
+
+func (st *StorageMarketActorState_I) _creditStorageDeals(rt Runtime, dealIDs []deal.DealID, lastChallengeEndEpoch block.ChainEpoch) {
+
+	for _, dealID := range dealIDs {
+		activeDeal, found := st.ActiveDeals()[dealID]
+
+		if !found {
+			rt.Abort("sma._creditStorageDeals: dealID not found.")
+		}
+
+		duration := lastChallengeEndEpoch - activeDeal.LastPaymentEpoch()
+
+		if duration <= 0 {
+			// return for this particular deal instead of aborting the whole operation
+			// rt.Abort("sma._creditStorageDeals: no new payment to be credited.")
+			return
+		}
+
+		fee := actor.TokenAmount(uint64(duration) * uint64(activeDeal.Deal().Proposal().StoragePricePerEpoch()))
+
+		if fee > activeDeal.LockedStorageFee() {
+			// rt.Abort("sma._creditStorageDeals: cannot unlock more than already locked.")
+			return
+		}
+
+		activeDeal.Impl().LastPaymentEpoch_ = lastChallengeEndEpoch
+
+		// potentially unnecessary as we can unlock funds for provider directly
+		// unless the protocol plans on refunding client for MaxFaultCount consecutive fails
+		activeDeal.Impl().UnlockStorageFee(fee)
+
+		// TODO: align on provider storage fee withdrawal
+	}
+
+}
+
+func (st *StorageMarketActorState_I) _slashDeclaredFaults(rt Runtime, dealIDs []deal.DealID) {
+
+	for _, dealID := range dealIDs {
+
+		deal, found := st.ActiveDeals()[dealID]
+
+		if !found {
+			// move on to the next deal and keep slashing
+			return
+		}
+
+		// TODO: the exact slash amount is up for change
+		// TODO: check if provider runs out of collateral to slash here
+		deal.Impl().ProviderCollateralRemaining_ -= deal.Deal().Proposal().ProviderCollateralPerEpoch()
+
+	}
+}
+
+func (st *StorageMarketActorState_I) _slashDetectedFaults(rt Runtime, dealIDs []deal.DealID) {
+
+	for _, dealID := range dealIDs {
+
+		deal, found := st.ActiveDeals()[dealID]
+
+		if !found {
+			// move on to the next deal and keep slashing
+			return
+		}
+
+		// TODO: the exact slash amount is up for change
+		// TODO: check if provider runs out of collateral to slash here
+		// TODO: more sever slashing for detected faults vs declared faults
+		// maybe for the duration since the last Epoch that the sector was proven
+		deal.Impl().ProviderCollateralRemaining_ -= deal.Deal().Proposal().ProviderCollateralPerEpoch()
+
+	}
+}
+
+// delete deal from active deals
+// send deal collateral to TreasuryActor
+// return locked storage fee to client
+// return client collateral
+// TODO: decide what to do with unlocked storage fee here
+func (st *StorageMarketActorState_I) _slashTerminatedFaults(rt Runtime, dealIDs []deal.DealID) {
+
+	for _, dealID := range dealIDs {
+
+		deal, found := st.ActiveDeals()[dealID]
+
+		if !found {
+			// move on to the next deal and keep slashing
+			return
+		}
+
+		delete(st.ActiveDeals(), dealID)
```

Comment:
- this should be in a function like `st._terminateDeal(dealID)` or `st._clearActiveDeal(dealID)` or something. stray deletes are asking for state-keeping trouble. make the transitions very explicit and clear, so state accounting is easy. 
- ie, here in this function we want to call a higher level logical construct ("a SlashTerminatedFaults messages also terminates the deal accounting -- therefore call deal termination, or deal cleanup")